crates/core/tedge_agent/src/state_repository/state.rs

```diff
@@ -18,8 +22,27 @@ pub fn agent_state_dir(tedge_root: Utf8PathBuf) -> Utf8PathBuf {
 }
 
 impl<T: DeserializeOwned + Serialize> AgentStateRepository<T> {
-    pub fn new(tedge_root: Utf8PathBuf, file_name: &str) -> Self {
-        let state_repo_path = agent_state_dir(tedge_root).join(file_name);
+    /// Create a new agent state file in the the given state directory.
+    ///
+    /// If for some reason the state file cannot be created (e.g. the directory doesn't exist or is not writable)
+    /// then the agent state is created under the tedge root directory (`/etc/tedge/.agent`).
+    pub fn new(state_dir: Utf8PathBuf, tedge_root: Utf8PathBuf, file_name: &str) -> Self {
+        // Check that the given directory is actually writable
+        let test_file = state_dir.join(state_dir.join(".--test--"));
+        let state_dir =
+            match File::create(test_file.clone()).and_then(|mut file| file.write_all(b"")) {
+                Ok(_) => {
+                    let _ = std::fs::remove_file(test_file);
+                    state_dir
+                }
+                Err(err) => {
+                    error!("Cannot use `{state_dir}` to store tedge-agent state: {err}");
```

Comment:
```suggestion
                    error!("Cannot use {state_dir:?} to store tedge-agent state: {err}");
```

I'm not entirely sure whether this is actually better, but it does stop the quotes being moved around incorrectly if an error message is changed in the future. Equally it's a pattern that does make it easy to possibly print things in ways we don't want to.